sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala

```diff
@@ -17,18 +17,270 @@
+   */
+  def remove(): UnsafeRow
+
+  /**
+   * Cleanup all the resources.
+   */
+  def close(): Unit
+}
+
+/**
+ * A RowQueue that is based on in-memory page. UnsafeRows are appended into it until it's full.
+ * Another thread could read from it at the same time (behind the writer).
+ */
+private[python] case class InMemoryRowQueue(page: MemoryBlock, fields: Int) extends RowQueue {
+  private val base: AnyRef = page.getBaseObject
+  private var last = page.getBaseOffset  // for writing
+  private var first = page.getBaseOffset  // for reading
+  private val resultRow = new UnsafeRow(fields)
+
+  def add(row: UnsafeRow): Boolean = {
+    if (last + 4 + row.getSizeInBytes > page.getBaseOffset + page.size) {
+      if (last + 4 <= page.getBaseOffset + page.size) {
+        Platform.putInt(base, last, -1)
+      }
+      return false
+    }
+    Platform.putInt(base, last, row.getSizeInBytes)
+    Platform.copyMemory(row.getBaseObject, row.getBaseOffset, base, last + 4, row.getSizeInBytes)
+    last += 4 + row.getSizeInBytes
+    true
+  }
+
+  def remove(): UnsafeRow = {
+    if (first + 4 > page.getBaseOffset + page.size || Platform.getInt(base, first) < 0) {
+      null
+    } else {
+      val size = Platform.getInt(base, first)
+      resultRow.pointTo(base, first + 4, size)
+      first += 4 + size
+      resultRow
+    }
+  }
+
+  def close(): Unit = {
+    // caller should override close() to free page
+  }
+}
+
+/**
+ * A RowQueue that is based on file in disk. It will stop to push once someone start to read
+ * from it.
+ */
+private[python] case class DiskRowQueue(path: String, fields: Int) extends RowQueue {
+  private var fout = new FileOutputStream(path)
+  private var out = new DataOutputStream(new BufferedOutputStream(fout))
+  private var length = 0L
+
+  private var fin: FileInputStream = _
+  private var in: DataInputStream = _
+  private val resultRow = new UnsafeRow(fields)
+
+  def add(row: UnsafeRow): Boolean = {
+    synchronized {
+      if (out == null) {
+        // Another thread is reading, stop writing this one
+        return false
+      }
+    }
+    out.writeInt(row.getSizeInBytes)
+    out.write(row.getBytes)
+    length += 4 + row.getSizeInBytes
+    true
+  }
+
+  def remove(): UnsafeRow = {
+    synchronized {
+      if (out != null) {
+        out.flush()
+        out.close()
+        out = null
+        fout.close()
+        fout = null
+
+        fin = new FileInputStream(path)
+        in = new DataInputStream(new BufferedInputStream(fin))
+      }
+    }
+
+    if (length > 0) {
+      val size = in.readInt()
+      assert(4 + size <= length, s"require ${4 + size} bytes for next row")
+      val bytes = new Array[Byte](size)
+      in.readFully(bytes)
+      length -= 4 + size
+      resultRow.pointTo(bytes, size)
+      resultRow
+    } else {
+      null
+    }
+  }
+
+  def close(): Unit = {
+    synchronized {
+      if (fout != null) {
```

Comment:
If you want to save a couple of lines, you could do

```scala
Closeables.close(fout, /* swallowIOException = */ true)
fout = null
Closeables.close(fin, /* swallowIOException = */ true)
fin = null
```

This also guards against IOExceptions in close since `Closeables` should log them.